modules/stochastic_tools/include/state_sim/model/actions/StateSimAction.h

```diff
@@ -0,0 +1,38 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef STATESIMACTION_H
+#define STATESIMACTION_H
+
+#include "StateSimBase.h"
+#include <string>
+
+class StateSimAction : public StateSimBase
+{
+protected:
+  ACTION_TYPE_ENUM _act_type;
+
+private:
+  static unsigned int _counter;
```

Comment:
Well an ID doesn't really exist like you are looking for. Consider a tree of multiapps. The indexing would have to be non-zero for all but the first branch.

I think a better idea would be to retrieve the ID from the Multiapp itself. Then you don't need atomics and then it would work with arbitrary nesting.